### PR TITLE
executor, planner: refine merge join compare methods to avoid some rare cases

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

fix merge join incorrect result in some rare cases.

### What is changed and how it works?

change the compare method

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -2268,16 +2268,41 @@ func resOfNE(val int64, isNull bool, err error) (int64, bool, error) {
...
+		return 0, true, err
+	}
+
+	arg1, isNull1, err := rhsArg.EvalInt(sctx, rhsRow)
+	if err != nil {
+		return 0, true, err
+	}
+
+	// compare null values.
+	if isNull0 || isNull1 {
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/builtin_compare.go]: no, it's to make it behave the same with `chunk.CompareFunc`

[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/builtin_compare.go]: i.e. NULL is considered to be equal to NULL and smaller than any non-NULL value

File: executor/merge_join.go [hunk 2]
```diff
@@ -340,6 +342,22 @@ func (e *MergeJoinExec) joinToChunk(ctx context.Context, chk *chunk.Chunk) (hasM
 	}
 }
 
+func (e *MergeJoinExec) compare(outerRow, innerRow chunk.Row) (int, error) {
+	outerJoinKeys := e.outerTable.keys
+	innerJoinKeys := e.innerTable.joinKeys
+	for i := range outerJoinKeys {
+		cmp, _, err := e.compareFuncs[i](e.ctx, outerJoinKeys[i], innerJoinKeys[i], outerRow, innerRow)
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: We do not check the null-value?

Comment (zz-jason) [reply executor/merge_join.go]: yes, I've refactored the `expression.CompareFunc`s to compare NULL values, compare result are stored in the first return value, thus the second return value only indicates whether there is a NULL value.

**Pull request merged**